File: android/sports/src/test/java/com/fitconnect/android/sports/progression/ProgressionEngineTest.kt
package com.fitconnect.android.sports.progression

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class ProgressionEngineTest {

    private fun working(
        reps: Int,
        target: Int,
        weight: Double = 60.0,
        failed: Boolean = false,
    ) = PreviousSetPerformance(
        setType = SetType.WORKING,
        actualReps = reps,
        actualWeightKg = weight,
        actualTimeSec = null,
        targetReps = target,
        targetWeightKg = weight,
        isFailed = failed,
    )

    @Test
    fun failedRepsDoNotAdvanceLinearLoad() {
        val input = ProgressionInput(
            rule = ProgressionRule.LINEAR,
            exerciseMode = ExerciseMode.REPS,
            sideMode = SideMode.NONE,
            previousSets = listOf(working(6, 8, failed = true)),
        )
        val target = ProgressionEngine.compute(input)
        assertEquals(60.0, target.targetWeightKg!!, 0.01)
        assertEquals(ProgressionState.HOLD, target.progressionState)
    }

    @Test
    fun stallTriggersDeload() {
        val input = ProgressionInput(
            rule = ProgressionRule.LINEAR,
            exerciseMode = ExerciseMode.REPS,
            sideMode = SideMode.NONE,
            previousSets = listOf(working(8, 8)),
            stallCount = 3,
        )
        val target = ProgressionEngine.compute(input)
        assertEquals(ProgressionState.DELOAD, target.progressionState)
        assertEquals(54.0, target.targetWeightKg!!, 0.01)
    }

    @Test
    fun bodyweightUsesRepProgression() {
        val input = ProgressionInput(
            rule = ProgressionRule.DOUBLE_PROGRESSION,
            exerciseMode = ExerciseMode.BODYWEIGHT,
            sideMode = SideMode.NONE,
            previousSets = listOf(working(10, 10, weight = 0.0)),
            repMin = 8,
            repMax = 12,
        )
        val target = ProgressionEngine.compute(input)
        assertNull(target.targetWeightKg)
        assertEquals(11, target.targetRepsMin)
    }

    @Test
    fun perSideRepsStayEven() {
        val input = ProgressionInput(
            rule = ProgressionRule.DOUBLE_PROGRESSION,
            exerciseMode = ExerciseMode.BODYWEIGHT,
            sideMode = SideMode.PER_SIDE,
            previousSets = listOf(working(8, 8, weight = 0.0)),
        )
        val target = ProgressionEngine.compute(input)
        assertEquals(0, target.targetRepsMin % 2)
    }
}

class OneRepMaxEstimatorTest {
    @Test
    fun estimateIncludesDisclaimer() {
        val est = OneRepMaxEstimator.estimate(100.0, 5)
        assertEquals(100.0, est.sourceWeightKg, 0.01)
        assertEquals(5, est.sourceReps)
        assert(est.disclaimer.contains("Estimated"))
        assert(est.estimatedKg > 100.0)
    }
}

File: apps/web/tests/strength/progression-engine.test.ts
import { describe, expect, it } from "vitest";
import { computeProgressionTarget, estimateOneRepMax } from "@fitconnect/utils";
import type { PreviousSetPerformance, ProgressionInput } from "@fitconnect/types";

const workingSet = (
  reps: number,
  target: number,
  weight = 60,
  failed = false
): PreviousSetPerformance => ({
  setType: "working",
  actualReps: reps,
  actualWeightKg: weight,
  actualTimeSec: null,
  targetReps: target,
  targetWeightKg: weight,
  isFailed: failed
});

describe("ProgressionEngine (TS)", () => {
  it("TEST 004 — failed reps do not advance linear load", () => {
    const input: ProgressionInput = {
      rule: "LINEAR",
      exerciseMode: "REPS",
      sideMode: "none",
      previousSets: [workingSet(6, 8, 60, true)],
      repMin: 6,
      repMax: 8,
      weightStepKg: 2.5,
      timeStepSec: 5,
      stallCount: 0,
      deloadPercent: 0.1
    };
    const target = computeProgressionTarget(input);
    expect(target.targetWeightKg).toBe(60);
    expect(target.progressionState).toBe("hold");
    expect(target.rationale).toContain("missed");
  });

  it("TEST 005 — stall triggers deload", () => {
    const input: ProgressionInput = {
      rule: "LINEAR",
      exerciseMode: "REPS",
      sideMode: "none",
      previousSets: [workingSet(8, 8)],
      repMin: 6,
      repMax: 8,
      weightStepKg: 2.5,
      timeStepSec: 5,
      stallCount: 3,
      deloadPercent: 0.1
    };
    const target = computeProgressionTarget(input);
    expect(target.progressionState).toBe("deload");
    expect(target.targetWeightKg).toBe(54);
  });

  it("TEST 008 — bodyweight uses rep progression not fake weight", () => {
    const input: ProgressionInput = {
      rule: "DOUBLE_PROGRESSION",
      exerciseMode: "BODYWEIGHT",
      sideMode: "none",
      previousSets: [workingSet(10, 10, 0)],
      repMin: 8,
      repMax: 12,
      weightStepKg: 0,
      timeStepSec: 0,
      stallCount: 0,
      deloadPercent: 0.1
    };
    const target = computeProgressionTarget(input);
    expect(target.targetWeightKg).toBeNull();
    expect(target.targetRepsMin).toBe(11);
  });

  it("TEST 009 — per-side reps stay even", () => {
    const input: ProgressionInput = {
      rule: "DOUBLE_PROGRESSION",
      exerciseMode: "BODYWEIGHT",
      sideMode: "per_side",
      previousSets: [workingSet(8, 8, 0)],
      repMin: 6,
      repMax: 10,
      weightStepKg: 0,
      timeStepSec: 0,
      stallCount: 0,
      deloadPercent: 0.1
    };
    const target = computeProgressionTarget(input);
    expect(target.targetRepsMin % 2).toBe(0);
  });

  it("double progression advances weight at top of range", () => {
    const input: ProgressionInput = {
      rule: "DOUBLE_PROGRESSION",
      exerciseMode: "REPS",
      sideMode: "none",
      previousSets: [workingSet(8, 8, 60)],
      repMin: 6,
      repMax: 8,
      weightStepKg: 2.5,
      timeStepSec: 5,
      stallCount: 0,
      deloadPercent: 0.1
    };
    const target = computeProgressionTarget(input);
    expect(target.targetWeightKg).toBe(62.5);
    expect(target.rationale).toContain("rep range");
  });
});

describe("OneRepMaxEstimator (TS)", () => {
  it("TEST 011 — returns source set and disclaimer", () => {
    const est = estimateOneRepMax(100, 5);
    expect(est.sourceWeightKg).toBe(100);
    expect(est.sourceReps).toBe(5);
    expect(est.disclaimer).toContain("Estimated");
    expect(est.estimatedKg).toBeGreaterThan(100);
  });
});
